### DISCUSSION TITLE
Date axis POSSIBLE?

### DISCUSSION THREAD

**Discussion by dirri** [GSI-CS-CO/chart-fx#521] [Q&A] [answered]
I need to add Date objects to XAxis as existings solution uses a custom DateAxis. Is it possible somehow with chartFx?
What is do i extract dates from a logfile and add for every finding the date on x axis and a value (ms) in Y axis. There are no fixed linear times as these events are non linear.

**Comment by wirew0rm** [comment, marked as answer]
Hello and welcome to chart-fx,
for performance reasons, our `DataSet` interface does not use Objects to store individual data points, since that would cause lots of allocations and indirections when rendering the data. Charts with time data are supported by the DefaultNumericAxis implementation, see the [TimeAxisSample](https://github.com/fair-acc/chart-fx/blob/master/chartfx-samples/src/main/java/de/gsi/chart/samples/TimeAxisSample.java). You will have to convert the `Date` objects to a <strike>milli</strike>second timestamp and add them to a suitable DataSet implementation. Most of chart-fx's DataSets do support non-equidistant samples.
For influencing how and where the dates will be printed on the x-axis, have a look at the different samples.

I hope this is helpful and I undersood your requirements correctly. In general we try not to actively prevent any use of chart-fx, so most things should be possible, the question is more how much custom code you need to get your desired result.

**Reply by dirri** [reply]
Thx for the reply. Yes thats what i needed to know. I´ve seen in the samples that they work with secs not millis...What DataSet would fit for milliseconds here? How does the implementation know the difference millis or seconds to interpret the date data and set the correct labels?

**Reply by wirew0rm** [reply]
Sorry, you are right, the axis uses second as the SI unit for time. If you implement the DataSet interface yourself, you can store the data in any format you like (even as date objects) as long as its `get(DIM_X, index)` method returns the timestamp in seconds. This dataset can still provide your own custom accessor to get the date in the format which you prefer.

The other avenue is to use a custom Axis, then you can interpret the data any way you want, but it will still be a double value and you will have to do a lot of the foot-work that `DefaultNumericAxis` provides yourself (but giving you absolute control over every tick mark).

**Reply by dirri** [reply]
I think i´ll use the default DefaultErrorDataSet and convert the millis to secs. But if i understand corrctly i need to know the start and endtime for XAxis when creating ist right? So i would need to eval first the latest timestamp and then create the Axis. Maybe i need to have a look onto the dynamic sample. What about timezone and format of the labels displayed in Timebased XAxis? Is it configurable?
Thx a lot for your help!!

**Comment by wirew0rm** [comment]
You can either set the axis range manually or manually set the axis range depending on your usecase. Looking at and playing with the samples is always a good idea there.
For changing the Timezone, see here: https://github.com/fair-acc/chart-fx/blob/0dd09be29c02d628adcc609086c04a0fb4f7fe6e/chartfx-chart/src/main/java/de/gsi/chart/axes/spi/format/DefaultTimeFormatter.java#L112

**Reply by dirri** [reply]
Thanks a lot i´m nearby..this is my chart using a custom DateAxis and standard XYChart:

![grafik](https://user-images.githubusercontent.com/11440450/169901063-d2f98562-d2ae-463d-8bd1-9994f9b77cbb.png)

And this is where i got so far seems timezone needs to be adapted and i need those nice circles as symbols ticks as in original :

![grafik](https://user-images.githubusercontent.com/11440450/169900923-016b4bed-3565-4318-a877-9263ba7aaf9b.png)

So as u see timestamps do not match exactly and i need the circle symbol ticks..btw what is the triangle drawn over the chart from beginning to the end? And i need to change the dataset name symbol how can i do that?

**Comment by dirri** [comment]
For compability reasons i´m using the JDK8 SNAPSHOT version of chartfx and maybe that is the reason why i can´t set the markerStyle to circle? I´m trying to set the style for the dataset (DoubleDataSet) via setStyle("strokeColor: red;strokeWidt: 3;markerType. circle;"); But this seems not to work. Are there any restrictions in the JDK8 SNAPSHOT version for styling the charts??

**Reply by wirew0rm** [reply]
Please note that the JDK8 version is not actively maintained anymore, since there is no publicly maintained version of JDK8 anymore. This is the cause of the spurious "triangle" lines across your data. You might be able to navigate around that bug by disabling data reduction.

Regarding marker styling, on first glance I see two typos in your css string: "Widt" without h and "markerType" needs to be followed by a colon. Otherwise try to set the marker style directly in the renderer: https://github.com/fair-acc/chart-fx/blob/0dd09be29c02d628adcc609086c04a0fb4f7fe6e/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/ErrorDataSetRenderer.java#L262

**Reply by dirri** [reply]
Yes the typos were not in code actually. The issue i now have is that i need to remove the GridRenderer from XYChart because its destroying my layout. It seems it kills my own renderer (to change the legendsymbol) and kills my styles added to dataset.
And although i cleanup the renderers list when i load at first time my chart the grids are visible in the first chart either. On reload they disappear. I like the grid but it does not help because the style of legendSymbol and dataset will be killed when not removing the Gridrenderer. Is this already known? Or how to prevent the Gridrenderer to shoot my other style changes?

**Reply by dirri** [reply]
To show what i mean, this is what comes up when not removing the GridRenderer it overwrites my changes coming from customrenderer and styling the dataset:
![grafik](https://user-images.githubusercontent.com/11440450/170187753-52ed892b-06b8-4c63-9646-1c3db020c9d4.png)


And this is what i get removing the Gridrenderer and what is supposed to look like :
![grafik](https://user-images.githubusercontent.com/11440450/170187827-f6f48802-0e9b-4677-9810-731529916716.png)

With "removing the Gridrenderer" i mean clear the renderers list before adding my own renderer for customizing the legend symbol. It seems the problem isn´t gridrenderer but the default ErrorDataSetRenderer.

**Reply by wirew0rm** [reply]
The Grid Renderer should not be in the RendererList, but is a separate property of the chart. Depending on how you initialize your chart and which version of chartfx you use, the renderer list is populated with a DefaultErrorDataSetRenderer. I usually use the `setAll(renderer1, renderer2, ...)` call on the renderer list whenever i use custom renderers.
I also think I remember that there where inconsistencies with legend smbols which are fixed on master but not in the last JDK8 release.

**Reply by dirri** [reply]
Sometimes when i zoom in by drawing a rectangle vis LMB over a certain area of the chart the result is an empty canvas. This seems to be sporadically. Also when using the mouse wheel result sometime to empty canvas. I did not found out yet what the reason is!?

**Reply by wirew0rm** [reply]
see reply here: https://github.com/fair-acc/chart-fx/discussions/524#discussioncomment-2897355
